Add documentation on using virtual environments for testing in napari based on 2022-01 workshop by Talley Lambert (#50)
napari/docs: docs/plugins/virtual_environment_docs/2-deploying-your-plugin.md

# 2: Deploying your plugin  

This guide explains some of the techniques you can use to deploy your plugin. 

## Other topics in this series:  
* 1: [Virtual environments](./1-virtual-environments.md)   
* 2: This guide  
* 3: [Version management](./3-version-management.md)     
* 4: [Developer tools](./4-developer-tools.md)   
* 5: [Survey](./5-Survey.md)   

## This guide covers:   
* [PyPI and Anaconda](#pypi-and-anaconda)
    - [Package formats (PyPI)](#package-formats-pypi)
    - [PyPI](#pypi)
* [Deploying plugins](#deploying-plugins)
    - [Automatically via GitHub actions](#automatically-via-github-actions)
    - [Anaconda cloud via conda-forge](#anaconda-cloud-via-conda-forge)

## PyPI and Anaconda

> Try to deploy to both! But for now, try to at least use pip. You can always also provide your users with manual installation instructions (e.g. if you want to use conda or have specific dependencies).

Reviewer: Would add just a short sentence here about the two:

```suggestion
PyPI and Anaconda are two options for how you distribute your package and allow your users to more easily find and install it.  Try to deploy to both! But for now, try to at least use PyPI. You can always also provide your users with manual installation instructions (e.g. if you want them to use `conda` or have specific dependencies).
```

### Package formats (PyPI)
**sdist** means source distribution. An **sdist** includes all of the files that are required to *build* your package. An **sdist** may require specific additional software (e.g. compilers) to actually build.

**wheel** is a prebuilt package, ready to drop into site-packages. It includes compiled OS-specific extensions (if applicable).

You are *strongly* encouraged to ship both! If the wheel is not present, pip will try to build it from the sdist before installation, and that may fail depending on the package. To see if a given package ships a wheel, check here: https://pypi.org/project/napari/#files

**Note:** This goes for dependencies too! Check all your dependencies for wheel availability.  

> ### PyPI

Reviewer: This section should be moved to `Deploying plugins` and it should be called `Manually via twine`.

Author: Done.

> **build** (package builder): https://pypa-build.readthedocs.io/en/latest/  

Reviewer: ```suggestion
**[build](https://pypa-build.readthedocs.io/en/latest/ )** : package builder that bundles your source code into `sdist` or `wheel` distributions
```

> **twine** (PyPI Client): https://twine.readthedocs.io/en/latest/  

Reviewer: ```suggestion
**[twine](https://twine.readthedocs.io/en/latest/ )**: client you can use to upload your distribution to PyPI. Note that you will need to set up a PyPI account and authenticate yourself when uploading. See [this great guide](https://packaging.python.org/en/latest/tutorials/packaging-projects/) for a detailed tutorial to building & sharing your first Python packages.
```

> ```python

Reviewer: ```suggestion
```console
```

    # twine is a PyPI Client  
    # build is a PEP 517 package builder  
    $ pip install twine build

    # create a wheel and an sdist  
    $ python -m build  
     
    # (Optional) upload to test PyPI  
    $ twine upload -r testpypi dist/*  

    # Upload to PyPI  
    $ twine upload dist/*  

```

> **Note:** `python -m build` is the modern alternative to setuptools `python setup.py sdist bdist_wheel`. (It calls setuptools behind the scenes.)

Reviewer: ```suggestion
**Note:** `python -m build` is the modern alternative to `setuptools`' `python setup.py sdist bdist_wheel`. It calls `setuptools` behind the scenes.
```

## Deploying plugins:

### Automatically via GitHub actions
This requires either: 

> * Running twine manually (e.g: your-repo/.github/workflows/deploy.yaml)  

Reviewer: ```suggestion
* Running `twine` as above in a workflow after setting up Python and installing it
```

or   

> * Using a pre-made GitHub action: https://github.com/pypa/gh-action-pypi-publish   

Reviewer: ```suggestion
* Using a pre-made [GitHub action](https://github.com/pypa/gh-action-pypi-publish)
```

Here is an example workflow that manually deploys using `twine` when tests pass and you push a tagged commit.

> ```yaml

Reviewer: ```suggestion
```yaml
#your-repo/.github/workflows/deploy.yaml
```

jobs:  
    deploy:  
      name: Deploy
      runs on: ubuntu-latest  
      if: “success() && startsWith(github.ref, ‘refs/tags/’)”  
      steps:  
    - uses: actions/checkout@v2  

    -     name: Set up Python  
           uses: actions/setup-python@v2  
     -     with:   
           python-version “3.x”  

     -     name: install  
           run:  
               pip install -U pip  
               pip install -U build twine  
               python -m build  
  
    -     name: Build and publish  
          run:  twine upload dist/*  
          env: 
             TWINE_USERNAME: __token__  
             TWINE_PASSWORD: ${{ secrets.TWINE_API_KEY }}  

```

>   **Note:** Gate this action on some criterion, like a git tag: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_Idif

Reviewer: This note makes more sense just after line 62 where we first mention a tagged commit. Also should be a hyperlink.

```suggestion
  **Note:** Gate this action on some criterion, like a [git tag](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_Idif).
```

# Deploying your plugin  

This guide explains some of the techniques you can use to deploy your plugin. 

> ## This guide covers:   

Reviewer: Hmm this structure is a little bit confusing. We only really talk about Anaconda in the very last bit, so I would make that its own heading, and focus on PyPI in the first two sections. So this would be my suggested structure:

```
* PyPI package formats
    - Overview
    - Building your package
* Deploying to PyPI
    - Manually via twine
    - Automatically via GitHub actions
* Deploying to Anaconda
```

Author: Good idea. I will change this.

Author: Ok, @DragaDoncila, please see what you think now.  It is not exactly like you suggested but I think it makes sense.

* [PyPI and Anaconda](#pypi-and-anaconda)
    - [Package formats (PyPI)](#package-formats-pypi)
    - [PyPI](#pypi)

> * [Deploying plugins](#deploying-plugins)

Reviewer: ```suggestion
* [Deploying plugins](#deploying-plugins)
    - [Manually via twine](#manually-via-twine)
```

    - [Automatically via GitHub actions](#automatically-via-github-actions)
* [Deploying plugins to Anaconda](#deploying-to-anaconda)

## Overview of PyPI and Anaconda
PyPI and Anaconda are two options for how you distribute your package and allow your users to more easily find and install it.  Try to deploy to both! But for now, try to at least use PyPI. You can always also provide your users with manual installation instructions (e.g. if you want them to use `conda` or have specific dependencies).

### Building your package 

> **sdist** means source distribution. An **sdist** includes all of the files that are required to *build* your package. An **sdist** may require specific additional software (e.g. compilers) to actually build.

Reviewer: ```suggestion
`sdist` means source distribution. An `sdist` includes all of the files that are required to *build* your package. An `sdist` may require specific additional software (e.g. compilers) to actually build.
```

I think we should swap all bolded references to `sdist` and `wheel` with backticks.

**wheel** is a prebuilt package, ready to drop into site-packages. It includes compiled OS-specific extensions (if applicable).

You are *strongly* encouraged to ship both! If the wheel is not present, pip will try to build it from the sdist before installation, and that may fail depending on the package. To see if a given package ships a wheel, check here: https://pypi.org/project/napari/#files

**Note:** This goes for dependencies too! Check all your dependencies for wheel availability.  

### PyPI

> **[build](https://pypa-build.readthedocs.io/en/latest/ )** : package builder that bundles your source code into `sdist` or `wheel` distributions

Reviewer: ```suggestion
**[build](https://pypa-build.readthedocs.io/en/latest/ )** is the recommended package builder that bundles your source code into `sdist` or `wheel` distributions. Install `build` into your local environment, and then run it at the root of your package to build your package:
```

> 

Reviewer: And the code block for the comment above...:

```console
pip install build
python -m build .
```

Author: I thought this was covered in ### Automatically via GitHub actions?

Reviewer: Building is slightly separate to deploying. You need to build to deploy, but I think we should have those two commands here to demonstrate the last sentence: _"Install `build` into your local environment and then run it at the root of your package to build your package."_

Author: Done.

## Deploying plugins:

### Manually via **twine**.  

> [twine](https://twine.readthedocs.io/en/latest/ )**: client you can use to upload your distribution to PyPI. Note that you will need to set up a PyPI account and authenticate yourself when uploading. See [this great guide](https://packaging.python.org/en/latest/tutorials/packaging-projects/) for a detailed tutorial to building and sharing your first Python packages.  

Reviewer: ```suggestion
[twine](https://twine.readthedocs.io/en/latest/ ) is a command line client you can use to upload your distribution to PyPI. Note that you will need to set up a PyPI account and authenticate yourself when uploading. See [this great guide](https://packaging.python.org/en/latest/tutorials/packaging-projects/) for a detailed tutorial to building and sharing your first Python packages.  
```

### Automatically via GitHub actions
This requires either: 
* Running `twine` as above in a workflow after setting up Python and installing it
or   
* Using a pre-made [GitHub action](https://github.com/pypa/gh-action-pypi-publish)

Here is an example workflow that manually deploys using `twine` when tests pass and you push a tagged commit.

```yaml
#your-repo/.github/workflows/deploy.yaml
jobs:  
    deploy:  
      name: Deploy
      runs on: ubuntu-latest  
      if: “success() && startsWith(github.ref, ‘refs/tags/’)”  
      steps:  
    - uses: actions/checkout@v2  

    -     name: Set up Python  
           uses: actions/setup-python@v2  
     -     with:   
           python-version “3.x”  

     -     name: install  
           run:  
               pip install -U pip  
               pip install -U build twine  
               python -m build  
  
    -     name: Build and publish  
          run:  twine upload dist/*  
          env: 
             TWINE_USERNAME: __token__  
             TWINE_PASSWORD: ${{ secrets.TWINE_API_KEY }}  

```

>   **Note:** Gate this action on some criterion, like a [git tag](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_Idif).

Reviewer: ```suggestion
  **Note:** Gate this action on some criterion, e.g. a git tag as above, or [some other criterion](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows).
```

### Anaconda cloud via conda-forge

> https://conda-forge.org/docs/maintainer/adding_pkgs.html

Reviewer: ```suggestion
This is only a brief guide to deploying to `conda-forge`. More information can be found in the [conda-forge docs](https://conda-forge.org/docs/maintainer/adding_pkgs.html).
```

1. Fork https://github.com/conda-forge/staged-recipes
2. Create a new branch
3. Create a new folder in recipes directory (copy the existing example)
4. Update the **meta.yaml** file to include your package's build and run requirements
5. Commit and open a PR to https://github.com/conda-forge/staged-recipes   

Once your recipe is approved and merged, the rest happens *automagically*, and your package will appear on the anaconda cloud.

> This is **far** easier if you already have an sdist published to PyPI.

Reviewer: ```suggestion
This is **far** easier if you already have an `sdist` published to PyPI.
```

## Other topics in this series:  
* [Virtual environments](./1-virtual-environments.md)   
* [Version management](./3-version-management.md)     
* [Developer tools](./4-developer-tools.md)   

> * [Survey/Q&A](./5-Survey.md)   

Reviewer: ```suggestion
* [Survey/Q&A](./5-survey.md)   
```

# Deploying your plugin  

This guide explains some of the techniques you can use to deploy your plugin. 

## This guide covers:   
* [Overview of PyPI and Anaconda](#overview-of-pypi-and-anaconda)
* [Building your package](#building-your-package)
* [Deploying plugins to PyPI](#deploying-plugins-to-pypi)
    - [Manually via twine](#manually-via-twine)
    - [Automatically via GitHub actions](#automatically-via-github-actions)
* [Deploying plugins to Anaconda](#deploying-to-anaconda)

## Overview of PyPI and Anaconda
PyPI and Anaconda are two options for how you distribute your package and allow your users to more easily find and install it.  Try to deploy to both! But for now, try to at least use PyPI. You can always also provide your users with manual installation instructions (e.g. if you want them to use `conda` or have specific dependencies).

### Building your package 
`sdist` means source distribution. An `sdist` includes all of the files that are required to *build* your package. An `sdist` may require specific additional software (e.g. compilers) to actually build.

> **wheel** is a prebuilt package, ready to drop into site-packages. It includes compiled OS-specific extensions (if applicable).

Reviewer: ```suggestion
`wheel` is a prebuilt package, ready to drop into your `site-packages` directory. It includes compiled OS-specific extensions (if applicable).
```

> You are *strongly* encouraged to ship both! If the wheel is not present, pip will try to build it from the sdist before installation, and that may fail depending on the package. To see if a given package ships a wheel, check here: https://pypi.org/project/napari/#files

Reviewer: ```suggestion
You are *strongly* encouraged to ship both! If the `wheel` is not present, `pip` will try to build it from the `sdist` before installation, and that may fail depending on the package. To see if a given package ships a `wheel`, check here: https://pypi.org/project/napari/#files
```

* Running `twine` as above in a workflow after setting up Python and installing it
or   
* Using a pre-made [GitHub action](https://github.com/pypa/gh-action-pypi-publish)

Here is an example workflow that manually deploys using `twine` when tests pass and you push a tagged commit.

```yaml
#your-repo/.github/workflows/deploy.yaml
jobs:  
    deploy:  
      name: Deploy
      runs on: ubuntu-latest  
      if: “success() && startsWith(github.ref, ‘refs/tags/’)”  
      steps:  
    - uses: actions/checkout@v2  

    -     name: Set up Python  
           uses: actions/setup-python@v2  
     -     with:   
           python-version “3.x”  

     -     name: install  
           run:  
               pip install -U pip  
               pip install -U build twine  
               python -m build  
  
    -     name: Build and publish  
          run:  twine upload dist/*  
          env: 
             TWINE_USERNAME: __token__  
             TWINE_PASSWORD: ${{ secrets.TWINE_API_KEY }}  

```

  **Note:** Gate this action on some criterion, e.g. a git tag as above, or [some other criterion](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows).

> ### Deploying to Anaconda

Reviewer: ```suggestion
## Deploying to Anaconda
```
